appengine/common/storage.js

```diff
@@ -60,71 +60,112 @@ BlocklyStorage.restoreBlocks = function() {
  */
 BlocklyStorage.link = function() {
   var code = BlocklyInterface.getCode();
-  BlocklyStorage.makeRequest_('/storage', 'xml', code);
+  BlocklyStorage.makeRequest('/storage', 'xml=' + encodeURIComponent(code),
+      BlocklyStorage.handleLinkResponse_);
 };
 
 /**
  * Retrieve XML text from database using given key.
  * @param {string} key Key to XML, obtained from href.
  */
 BlocklyStorage.retrieveXml = function(key) {
-  BlocklyStorage.makeRequest_('/storage', 'key', key);
+  BlocklyStorage.makeRequest('/storage', 'key=' + encodeURIComponent(key),
+      BlocklyStorage.handleRetrieveXmlResponse_);
 };
 
 /**
- * Global reference to current AJAX request.
- * @type XMLHttpRequest
- * @private
+ * Global reference to current AJAX requests.
+ * @type Object.<string, XMLHttpRequest>
  */
-BlocklyStorage.xhr_ = null;
+BlocklyStorage.xhr_ = {};
 
 /**
  * Fire a new AJAX request.
  * @param {string} url URL to fetch.
- * @param {string} name Name of parameter.
- * @param {string} content Content of parameter.
- * @private
+ * @param {Array.<string>} data Body of data to be sent in request.
+ * @param {?Function=} opt_onSuccess Function to call after request completes
+ *    successfully.
+ * @param {?Function=} opt_onFailure Function to call after request completes
+ *    unsuccessfully. Defaults to BlocklyStorage alert of request status.
+ * @param {string=} [opt_method='POST'] The HTTP request method to use.
+ */
+BlocklyStorage.makeFormRequest =
+    function(form, opt_onSuccess, opt_onFailure, opt_method) {
+  var data = [];
+  for (var i = 0, element; (element = form.elements[i]); i++) {
+    if (element.name) {
+      data[i] = encodeURIComponent(element.name) + '=' +
```

Comment:
This should be .push() since we might skip an element.

Reply:
Good point.